### ISSUE TITLE
Unused code generated when creating arrays in memory

### ISSUE THREAD

**Issue by winterkasten** [ethereum/solidity#3765]
Im currently running version 0.4.21 of solc on my linux platform. When compilig this code 
```solidity
pragma solidity ^0.4.0;

contract ArrayAccessV1 {
    uint x;

    function setStorageItem() external {
        uint[3] memory a = [uint(1),2,3];
        x = a[1];
    }
}
```
with optimization to evm assembly language, I recognized some instructions, which seems to me as unused. At the end of the assembly file there are two jump addresses:
```
tag_7:
      0x60
      mload(0x40)
      swap1
      dup2
      add
      0x40 
      mstore
      0x3
      dup2
tag_10:
      0x0
      dup2
      mstore
      0x20
      add
      swap1
      0x1
      swap1
      sub
      swap1
      dup2
      tag_10
      jumpi
      swap1
      pop
      pop
      swap1
      jump	// out
```
which are never jumped to during execution. These adresses are never placed on the stack for jumping. The only appearance of tag_7 beside this is in a commentary:
```
tag_5:
        /* "ArrayAccessV1.sol":112:128  uint[3] memory a */
      tag_6
      jump	// in(tag_7)
```
The instruction before tag_7 perform a unconditional jump to the stop instruction. So I think these basic blocks are unused and can be removed from the assembly. Or am I missing something?

**Comment by chriseth** [comment]
We disabled the unused code remover some time ago because removing blocks that start with tags that are not referenced can lead to invalid code. There are two situations where this can happen:

1. inline assembly that uses things like `jump(x)` where `x` is a variable
2. internal function identifiers stored in storage, especially when used inside the constructor

The first situation can only happen in deprecated code and we should certainly revisit the second one.

I created an issue for that: https://github.com/ethereum/solidity/issues/3766
